File: Plugins/RomanPort.ViewRdsBar/main/sdk/web/dispatchers/RaptorDispatcherOpcode.ts
import RaptorEventDispaptcher from "../../RaptorEventDispatcher";
import IRaptorEndpoint from "../IRaptorEndpoint";

export default class RaptorDispatcherOpcode {

    constructor(parent: IRaptorEndpoint) {
        this.parent = parent;
        this.subscriptions = [];
    }

    private parent: IRaptorEndpoint;
    private subscriptions: RaptorDispatcherOpcode_Subscription[];

    CreateSubscription(opcode: string) {
        var sub = new RaptorDispatcherOpcode_Subscription(this.parent, opcode);
        this.subscriptions.push(sub);
        return sub;
    }

}

class RaptorDispatcherOpcode_Subscription implements IRaptorEndpoint {

    constructor(parent: IRaptorEndpoint, opcode: string) {
        this.parent = parent;
        this.opcode = opcode;
        this.OnMessage = new RaptorEventDispaptcher<any>();
        parent.OnMessage.Bind((message: any) => {
            if (message["op"] == this.opcode) {
                this.OnMessage.Fire(message["d"]);
            }
        });
    }

    SendMessage(message: any): void {
        this.parent.SendMessage({
            "op": this.opcode,
            "d": message
        });
    }

    OnMessage: RaptorEventDispaptcher<any>;

    private parent: IRaptorEndpoint;
    private opcode: string;

}

File: RaptorSDR.Server/RaptorSDR.Server.Common/Dispatchers/RaptorDispatcherOpcode.cs
using Newtonsoft.Json.Linq;
using System;
using System.Collections.Generic;
using System.Text;

namespace RaptorSDR.Server.Common.Dispatchers
{
    public class RaptorDispatcherOpcode
    {
        private List<RaptorDispatcherOpcodeSubscription> subscriptions = new List<RaptorDispatcherOpcodeSubscription>();

        public RaptorDispatcherOpcode(IRaptorEndpoint parent)
        {
            this.parent = parent;
        }

        private IRaptorEndpoint parent;

        public IRaptorEndpoint CreateSubscription(string opcode)
        {
            RaptorDispatcherOpcodeSubscription subscription = new RaptorDispatcherOpcodeSubscription(parent, opcode);
            subscriptions.Add(subscription);
            return subscription;
        }

        class RaptorDispatcherOpcodeSubscription : IRaptorEndpoint
        {
            public RaptorDispatcherOpcodeSubscription(IRaptorEndpoint parent, string opcode)
            {
                this.parent = parent;
                this.opcode = opcode;
                parent.OnMessage += Parent_OnMessage;
                parent.OnClientLost += (IRaptorEndpoint ep, IRaptorEndpointClient client, IRaptorSession session) => OnClientLost?.Invoke(ep, client, session);
                parent.OnClientConnected += (IRaptorEndpoint ep, IRaptorEndpointClient client, IRaptorSession session) => OnClientConnected?.Invoke(ep, client, session);
            }

            private string opcode;
            private IRaptorEndpoint parent;

            public event IRaptorEndpoint_ClientStatusChanged OnClientLost;
            public event IRaptorEndpoint_ClientStatusChanged OnClientConnected;
            public event IRaptorEndpoint_MessageEventArgs OnMessage;

            private void Parent_OnMessage(IRaptorEndpoint ep, IRaptorEndpointClient client, JObject payload)
            {
                if (payload.ContainsKey("op") && payload.ContainsKey("d") && (string)payload["op"] == opcode)
                    OnMessage?.Invoke(this, client, (JObject)payload["d"]);
            }

            public void SendAll(JObject payload)
            {
                JObject output = new JObject();
                output["op"] = opcode;
                output["d"] = payload;
                parent.SendAll(output);
            }

            public void SendTo(IRaptorEndpointClient client, JObject payload)
            {
                JObject output = new JObject();
                output["op"] = opcode;
                output["d"] = payload;
                parent.SendTo(client, output);
            }

            public IRaptorEndpoint BindOnMessage(IRaptorEndpoint_MessageEventArgs evt)
            {
                OnMessage += evt;
                return this;
            }
        }
    }
}
